frontend/src/metabase/visualizations/echarts/cartesian/model/dataset.ts

```diff
@@ -590,27 +591,33 @@ export const applyVisualizationSettingsDataTransformations = (
 
 export const sortDataset = (
   dataset: ChartDataset,
-  xAxisScale?: XAxisScale,
+  xAxisScale: XAxisScale | undefined,
+  showWarning?: ShowWarning,
```

Comment:
I think this approach is not the optimal one because we are mixing something completely unrelated: showing warning logic into functions that have a different responsibility. Wdyt about changing it to the approach the RowChart has taken? It is separated there

Reply:
The RowChart is not a great comparison because it's implementation is very different, plus it only implements one out of the 5 warnings that are needed. This approach that I'm currently taking is close to the old dc.js approach, which implemented all warnings.

This PR is the third in a stack of four PRs (you can find the stack order in the automated comment). In the other PRs it's more apparent why it's better not to separate the logic.

In https://github.com/metabase/metabase/pull/40758, we need to compute the `seriesKey` of each column in order to determine if there are unaggregated values on that column. Doing this is not trivial, it requires using the `breakoutIndex` in case there is a breakout, which requires computation of the column descriptors.

https://github.com/metabase/metabase/pull/40759 requires dimension data to be in the dataset under the `X_AXIS_DATA_KEY`. We could circumvent this by using the dimension column directly, but again that requires computation of the column descriptors.

It's technically possible to redo these computations in a separate function just for getting warnings, but I don't think it's better because now a lot of logic has to be duplicated and we would have to manually make sure they are the same.

Reply:
Yes, the old dc.js code did it it this way but it is not great by design: we had pure functions that have one specific responsibility and then we inject a "showWarning" function which supposes to change some other state of application. For example, the function that is responsible for sorting the dataset reports about non-timeseries values. That is the reason the why a newer RowChart did not follow this approach although there are fewer cases to handle.

I think it is still reasonable to improve this aspect and compute warnings from chartModel and rawSeries separately. But if this takes too long, feel free to create a tech debt task for this and merge it as is